share/man/man8/vmstat.8
'\" te
.\"  Copyright (c) 2001 Sun Microsystems, Inc.  All Rights Reserved.
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License. You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.
.\"  See the License for the specific language governing permissions and limitations under the License. When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with
.\" the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH VMSTAT 8 "April 14, 2015"
.SH NAME
vmstat \- report virtual memory statistics
.SH SYNOPSIS
.LP
.nf
\fBvmstat\fR [\fB-ipqsS\fR] [\fB-T\fR u | d] [\fIdisks\fR] [\fIinterval\fR [\fIcount\fR]]
.fi

.SH DESCRIPTION
.LP
\fBvmstat\fR reports virtual memory statistics regarding kernel thread, virtual
memory, disk, trap, and \fBCPU\fR activity.
.sp
.LP
On \fBMP\fR (multi-processor) systems, \fBvmstat\fR averages the number of
\fBCPUs\fR into the output. For per-processor statistics, see \fBmpstat\fR(8).
.sp
.LP
\fBvmstat\fR only supports statistics for certain devices. For more general
system statistics, use \fBsar\fR(1), \fBiostat\fR(8), or \fBsar\fR(8).
.sp
.LP
Without options, \fBvmstat\fR displays a one-line summary of the virtual memory
activity since the system was booted.
.sp
.LP
During execution of the kernel status command, the \fBstate\fR of the system
can change. If relevant, a state change message is included in the \fBvmstat\fR
output, in one of the following forms:
.sp
.in +2
.nf
<<device added: sd0>>
<<device removed: sd0>>
<<processors added: 1, 3>>
<<processors removed: 1, 3>>
.fi
.in -2
.sp

.sp
.LP
See \fI\fR for device naming conventions for disks.
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-i\fR\fR
.ad
.RS 13n
Report the number of interrupts per device. \fIcount\fR and \fIinterval\fR does
not apply to the \fB-i\fR option.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR\fR
.ad
.RS 13n
Report paging activity in details. This option will display the following,
respectively:
.sp
.ne 2
.na
\fBepi\fR
.ad
.RS 7n
Executable page-ins.
.RE

.sp
.ne 2
.na
\fBepo\fR
.ad
.RS 7n
Executable page-outs.
.RE

.sp
.ne 2
.na
\fBepf\fR
.ad
.RS 7n
Executable page-frees.
.RE

.sp
.ne 2
.na
\fBapi\fR
.ad
.RS 7n
Anonymous page-ins.
.RE

.sp
.ne 2
.na
\fBapo\fR
.ad
.RS 7n
Anonymous page-outs.
.RE

.sp
.ne 2
.na
\fBapf\fR
.ad
.RS 7n
Anonymous page-frees.
.RE

.sp
.ne 2
.na
\fBfpi\fR
.ad
.RS 7n
File system page-ins.
.RE

.sp
.ne 2
.na
\fBfpo\fR
.ad
.RS 7n
File system page-outs.
.RE

.sp
.ne 2
.na
\fBfpf\fR
.ad
.RS 7n
File system page-frees.
.RE

When executed in a \fBzone\fR and if the pools facility is active, all of the
above only report actitivity on the processors in the processor set of the
\fBzone\fR's pool.
.RE

.sp
.ne 2
.na
\fB\fB-q\fR\fR
.ad
.RS 13n
Suppress messages related to state changes.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR\fR
.ad
.RS 13n
Display the total number of various system events since boot. \fIcount\fR and
\fIinterval\fR does not apply to the \fB-s\fR option.
.RE

.sp
.ne 2
.na
\fB\fB-S\fR\fR
.ad
.RS 13n
Report on swapping rather than paging activity. This option will change two
fields in \fBvmstat\fR's ``paging'' display: rather than the ``re'' and ``mf''
fields, \fBvmstat\fR will report ``si'' (swap-ins) and ``so'' (swap-outs).
.RE

.sp
.ne 2
.na
\fB\fB-T\fR \fBu\fR |  \fBd\fR\fR
.ad
.RS 13n
Specify \fBu\fR for a printed representation of the internal representation of
time. See \fBtime\fR(2). Specify \fBd\fR for standard date format. See
\fBdate\fR(1).
.RE

.SH OPERANDS
.LP
The following operands are supported:
.sp
.ne 2
.na
\fB\fIcount\fR\fR
.ad
.RS 12n
Specifies the number of times that the statistics are repeated. \fIcount\fR
does not apply to the \fB-i\fR and \fB-s\fR options.
.RE

.sp
.ne 2
.na
\fB\fIdisks\fR\fR
.ad
.RS 12n
Specifies which disks are to be given priority in the output (only four disks
fit on a line). Common disk names are \fBid\fR, \fBsd\fR, \fBxd\fR, or
\fBxy,\fR followed by a number (for example, \fBsd2\fR, \fBxd0\fR, and so
forth).
.RE

.sp
.ne 2
.na
\fB\fIinterval\fR\fR
.ad
.RS 12n
Specifies the last number of seconds over which \fBvmstat\fR summarizes
activity. This number of seconds repeats forever. \fIinterval\fR does not apply
to the \fB-i\fR and \fB-s\fR options.
.RE

.SH EXAMPLES
.LP
\fBExample 1 \fRUsing \fBvmstat\fR
.sp
.LP
The following command displays a summary of what the system is doing every five
seconds.

.sp
.in +2
.nf
example% \fBvmstat 5\fR


kthr   memory          page             disk      faults        cpu
r b w swap  free re mf pi p fr de sr s0 s1 s2 s3  in  sy  cs us sy id
0 0 0 11456 4120 1  41 19 1  3  0  2  0  4  0  0  48 112 130  4 14 82
0 0 1 10132 4280 0   4 44 0  0  0  0  0 23  0  0 211 230 144  3 35 62
0 0 1 10132 4616 0   0 20 0  0  0  0  0 19  0  0 150 172 146  3 33 64
0 0 1 10132 5292 0   0  9 0  0  0  0  0 21  0  0 165 105 130  1 21 78
1 1 1 10132 5496 0   0  5 0  0  0  0  0 23  0  0 183  92 134  1 20 79
1 0 1 10132 5564 0   0 25 0  0  0  0  0 18  0  0 131 231 116  4 34 62
1 0 1 10124 5412 0   0 37 0  0  0  0  0 22  0  0 166 179 118  1 33 67
1 0 1 10124 5236 0   0 24 0  0  0  0  0 14  0  0 109 243 113  4 56 39
^C

example%
.fi
.in -2
.sp

.sp
.LP
The fields of \fBvmstat\fR's display are

.sp
.ne 2
.na
\fB\fBkthr\fR\fR
.ad
.RS 10n
Report the number of kernel threads in each of the three following states:
.sp
.ne 2
.na
\fB\fBr\fR\fR
.ad
.RS 5n
the number of kernel threads in run queue
.RE

.sp
.ne 2
.na
\fB\fBb\fR\fR
.ad
.RS 5n
the number of blocked kernel threads that are waiting for resources \fBI/O,\fR
paging, and so forth
.RE

.sp
.ne 2
.na
\fB\fBw\fR\fR
.ad
.RS 5n
the number of swapped out lightweight processes (LWPs) that are waiting for
processing resources to finish.
.RE

.RE

.sp
.ne 2
.na
\fB\fBmemory\fR\fR
.ad
.RS 10n
Report on usage of virtual and real memory.
.sp
.ne 2
.na
\fB\fBswap\fR\fR
.ad
.RS 8n
available swap space (Kbytes)
.RE

.sp
.ne 2
.na
\fB\fBfree\fR\fR
.ad
.RS 8n
size of the free list (Kbytes)
.RE

.RE

.sp
.ne 2
.na
\fB\fBpage\fR\fR
.ad
.RS 10n
Report information about page faults and paging activity. The information on
each of the following activities is given in units per second.
.sp
.ne 2
.na
\fB\fBre\fR\fR
.ad
.RS 6n
page reclaims \(em but see the \fB-S\fR option for how this field is modified.
.RE

.sp
.ne 2
.na
\fB\fBmf\fR\fR
.ad
.RS 6n
minor faults \(em but see the \fB-S\fR option for how this field is modified.
.RE

.sp
.ne 2
.na
\fB\fBpi\fR\fR
.ad
.RS 6n
kilobytes paged in
.RE

.sp
.ne 2
.na
\fB\fBpo\fR\fR
.ad
.RS 6n
kilobytes paged out
.RE

.sp
.ne 2
.na
\fB\fBfr\fR\fR
.ad
.RS 6n
kilobytes freed
.RE

.sp
.ne 2
.na
\fB\fBde\fR\fR
.ad
.RS 6n
anticipated short-term memory shortfall (Kbytes)
.RE

.sp
.ne 2
.na
\fB\fBsr\fR\fR
.ad
.RS 6n
pages scanned by clock algorithm
.RE

When executed in a \fBzone\fR and if the pools facility is active, all of the
above (except for "de") only report activity on the processors in the processor
set of the \fBzone\fR's pool.
.RE

.sp
.ne 2
.na
\fB\fBdisk\fR\fR
.ad
.RS 10n
Report the number of disk operations per second. There are slots for up to four
disks, labeled with a single letter and number. The letter indicates the type
of disk (s = \fBSCSI\fR, i = \fBIPI\fR, and so forth); the number is the
logical unit number.
.RE

.sp
.ne 2
.na
\fB\fBfaults\fR\fR
.ad
.RS 10n
Report the trap/interrupt rates (per second).
.sp
.ne 2
.na
\fB\fBin\fR\fR
.ad
.RS 6n
interrupts
.RE

.sp
.ne 2
.na
\fB\fBsy\fR\fR
.ad
.RS 6n
system calls
.RE

.sp
.ne 2
.na
\fB\fBcs\fR\fR
.ad
.RS 6n
\fBCPU\fR context switches
.RE

When executed in a \fBzone\fR and if the pools facility is active, all of the
above only report actitivity on the processors in the processor set of the
\fBzone\fR's pool.
.RE

.sp
.ne 2
.na
\fB\fBcpu\fR\fR
.ad
.RS 10n
Give a breakdown of percentage usage of \fBCPU\fR time. On \fBMP\fR systems,
this is an average across all processors.
.sp
.ne 2
.na
\fB\fBus\fR\fR
.ad
.RS 6n
user time
.RE

.sp
.ne 2
.na
\fB\fBsy\fR\fR
.ad
.RS 6n
system time
.RE

.sp
.ne 2
.na
\fB\fBid\fR\fR
.ad
.RS 6n
idle time
.RE

When executed in a \fBzone\fR and if the pools facility is active, all of the
above only report actitivity on the processors in the processor set of the
\fBzone\fR's pool.
.RE

.SH ATTRIBUTES
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	See below.
.TE

.sp
.LP
Invocation is evolving. Human readable output is unstable.
.SH SEE ALSO
.LP
\fBdate\fR(1), \fBsar\fR(1), \fBiostat\fR(8), \fBmpstat\fR(8), \fBsar\fR(8),
\fBtime\fR(2), \fBattributes\fR(5)
.SH NOTES
.LP
The sum of CPU utilization might vary slightly from 100 because of rounding
errors in the production of a percentage figure.
.sp


usr/src/cmd/stat/vmstat/vmstat.c
/*
 * Copyright (c) 1991, 2010, Oracle and/or its affiliates. All rights reserved.
 */

/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

/* from UCB 5.4 5/17/86 */
/* from SunOS 4.1, SID 1.31 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <ctype.h>
#include <unistd.h>
#include <memory.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <values.h>
#include <poll.h>
#include <locale.h>

#include "statcommon.h"

char *cmdname = "vmstat";
int caught_cont = 0;

static uint_t timestamp_fmt = NODATE;

static	int	hz;
static	int	pagesize;
static	double	etime;
static	int	lines = 1;
static	int	swflag = 0, pflag = 0;
static	int	suppress_state;
static	long	iter = 0;
static	hrtime_t period_n = 0;
static  struct	snapshot *ss;

struct iodev_filter df;

#define	pgtok(a) ((a) * (pagesize >> 10))
#define	denom(x) ((x) ? (x) : 1)
#define	REPRINT	19

static	void	dovmstats(struct snapshot *old, struct snapshot *new);
static	void	printhdr(int);
static	void	dosum(struct sys_snapshot *ss);
static	void	dointr(struct snapshot *ss);
static	void	usage(void);

int
main(int argc, char **argv)
{
	struct snapshot *old = NULL;
	enum snapshot_types types = SNAP_SYSTEM;
	int summary = 0;
	int intr = 0;
	kstat_ctl_t *kc;
	int forever = 0;
	hrtime_t start_n;
	int c;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)		/* Should be defined by cc -D */
#define	TEXT_DOMAIN "SYS_TEST"		/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	pagesize = sysconf(_SC_PAGESIZE);
	hz = sysconf(_SC_CLK_TCK);

	while ((c = getopt(argc, argv, "ipqsST:")) != EOF)
		switch (c) {
		case 'S':
			swflag = !swflag;
			break;
		case 's':
			summary = 1;
			break;
		case 'i':
			intr = 1;
			break;
		case 'q':
			suppress_state = 1;
			break;
		case 'p':
			pflag++;	/* detailed paging info */
			break;
		case 'T':
			if (optarg) {
				if (*optarg == 'u')
					timestamp_fmt = UDATE;
				else if (*optarg == 'd')
					timestamp_fmt = DDATE;
				else
					usage();
			} else {
				usage();
			}
			break;
		default:
			usage();
		}

	argc -= optind;
	argv += optind;

	/* consistency with iostat */
	types |= SNAP_CPUS;

	if (intr)
		types |= SNAP_INTERRUPTS;
	if (!intr)
		types |= SNAP_IODEVS;

	/* max to fit in less than 80 characters */
	df.if_max_iodevs = 4;
	df.if_allowed_types = IODEV_DISK;
	df.if_nr_names = 0;
	df.if_names = safe_alloc(df.if_max_iodevs * sizeof (char *));
	(void) memset(df.if_names, 0, df.if_max_iodevs * sizeof (char *));

	while (argc > 0 && !isdigit(argv[0][0]) &&
	    df.if_nr_names < df.if_max_iodevs) {
		df.if_names[df.if_nr_names] = *argv;
		df.if_nr_names++;
		argc--, argv++;
	}

	kc = open_kstat();

	start_n = gethrtime();

	ss = acquire_snapshot(kc, types, &df);

	/* time, in seconds, since boot */
	etime = ss->s_sys.ss_ticks / hz;

	if (intr) {
		dointr(ss);
		free_snapshot(ss);
		exit(0);
	}
	if (summary) {
		dosum(&ss->s_sys);
		free_snapshot(ss);
		exit(0);
	}

	if (argc > 0) {
		long interval;
		char *endptr;

		errno = 0;
		interval = strtol(argv[0], &endptr, 10);

		if (errno > 0 || *endptr != '\0' || interval <= 0 ||
		    interval > MAXINT)
			usage();
		period_n = (hrtime_t)interval * NANOSEC;
		if (period_n <= 0)
			usage();
		iter = MAXLONG;
		if (argc > 1) {
			iter = strtol(argv[1], NULL, 10);
			if (errno > 0 || *endptr != '\0' || iter <= 0)
				usage();
		} else
			forever = 1;
		if (argc > 2)
			usage();
	}

	(void) sigset(SIGCONT, printhdr);

	dovmstats(old, ss);
	while (forever || --iter > 0) {
		/* (void) poll(NULL, 0, poll_interval); */

		/* Have a kip */
		sleep_until(&start_n, period_n, forever, &caught_cont);

		free_snapshot(old);
		old = ss;
		ss = acquire_snapshot(kc, types, &df);

		if (!suppress_state)
			snapshot_report_changes(old, ss);

		/* if config changed, show stats from boot */
		if (snapshot_has_changed(old, ss)) {
			free_snapshot(old);
			old = NULL;
		}

		dovmstats(old, ss);
	}

	free_snapshot(old);
	free_snapshot(ss);
	free(df.if_names);
	(void) kstat_close(kc);
	return (0);
}

#define	DELTA(v) (new->v - (old ? old->v : 0))
#define	ADJ(n)	((adj <= 0) ? n : (adj >= n) ? 1 : n - adj)
#define	adjprintf(fmt, n, val)	adj -= (n + 1) - printf(fmt, ADJ(n), val)

static int adj;	/* number of excess columns */

/*ARGSUSED*/
static void
show_disk(void *v1, void *v2, void *d)
{
	struct iodev_snapshot *old = (struct iodev_snapshot *)v1;
	struct iodev_snapshot *new = (struct iodev_snapshot *)v2;
	hrtime_t oldtime = new->is_crtime;
	double hr_etime;
	double reads, writes;

	if (new == NULL)
		return;

	if (old)
		oldtime = old->is_stats.wlastupdate;
	hr_etime = new->is_stats.wlastupdate - oldtime;
	if (hr_etime == 0.0)
		hr_etime = NANOSEC;
	reads = new->is_stats.reads - (old ? old->is_stats.reads : 0);
	writes = new->is_stats.writes - (old ? old->is_stats.writes : 0);
	adjprintf(" %*.0f", 2, (reads + writes) / hr_etime * NANOSEC);
}

static void
dovmstats(struct snapshot *old, struct snapshot *new)
{
	kstat_t *oldsys = NULL;
	kstat_t *newsys = &new->s_sys.ss_agg_sys;
	kstat_t *oldvm = NULL;
	kstat_t *newvm = &new->s_sys.ss_agg_vm;
	double percent_factor;
	ulong_t sys_updates, vm_updates;
	int count;

	adj = 0;

	if (old) {
		oldsys = &old->s_sys.ss_agg_sys;
		oldvm = &old->s_sys.ss_agg_vm;
	}

	etime = cpu_ticks_delta(oldsys, newsys);

	percent_factor = 100.0 / denom(etime);
	/*
	 * If any time has passed, convert etime to seconds per CPU
	 */
	etime = etime >= 1.0 ? (etime / nr_active_cpus(new)) / hz : 1.0;
	sys_updates = denom(DELTA(s_sys.ss_sysinfo.updates));
	vm_updates = denom(DELTA(s_sys.ss_vminfo.updates));

	if (timestamp_fmt != NODATE) {
		print_timestamp(timestamp_fmt);
		lines--;
	}

	if (--lines <= 0)
		printhdr(0);

	adj = 0;

	if (pflag) {
		adjprintf(" %*u", 6,
		    pgtok((int)(DELTA(s_sys.ss_vminfo.swap_avail)
		    / vm_updates)));
		adjprintf(" %*u", 5,
		    pgtok((int)(DELTA(s_sys.ss_vminfo.freemem) / vm_updates)));
		adjprintf(" %*.0f", 3, kstat_delta(oldvm, newvm, "pgrec")
		    / etime);
		adjprintf(" %*.0f", 3, (kstat_delta(oldvm, newvm, "hat_fault") +
		    kstat_delta(oldvm, newvm, "as_fault")) / etime);
		adjprintf(" %*.0f", 3, pgtok(kstat_delta(oldvm, newvm, "dfree"))
		    / etime);
		adjprintf(" %*ld", 3, pgtok(new->s_sys.ss_deficit));
		adjprintf(" %*.0f", 3, kstat_delta(oldvm, newvm, "scan")
		    / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "execpgin")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "execpgout")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "execfree")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "anonpgin")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "anonpgout")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "anonfree")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "fspgin")) / etime);
		adjprintf(" %*.0f", 4,
		    pgtok(kstat_delta(oldvm, newvm, "fspgout")) / etime);
		adjprintf(" %*.0f\n", 4,
		    pgtok(kstat_delta(oldvm, newvm, "fsfree")) / etime);
		(void) fflush(stdout);
		return;
	}

	adjprintf(" %*lu", 1, DELTA(s_sys.ss_sysinfo.runque) / sys_updates);
	adjprintf(" %*lu", 1, DELTA(s_sys.ss_sysinfo.waiting) / sys_updates);
	adjprintf(" %*lu", 1, DELTA(s_sys.ss_sysinfo.swpque) / sys_updates);
	adjprintf(" %*u", 6, pgtok((int)(DELTA(s_sys.ss_vminfo.swap_avail)
	    / vm_updates)));
	adjprintf(" %*u", 5, pgtok((int)(DELTA(s_sys.ss_vminfo.freemem)
	    / vm_updates)));
	adjprintf(" %*.0f", 3, swflag? 0 :
	    kstat_delta(oldvm, newvm, "pgrec") / etime);
	adjprintf(" %*.0f", 3, swflag? 0 :
	    (kstat_delta(oldvm, newvm, "hat_fault")
	    + kstat_delta(oldvm, newvm, "as_fault"))
	    / etime);
	adjprintf(" %*.0f", 2, pgtok(kstat_delta(oldvm, newvm, "pgpgin"))
	    / etime);
	adjprintf(" %*.0f", 2, pgtok(kstat_delta(oldvm, newvm, "pgpgout"))
	    / etime);
	adjprintf(" %*.0f", 2, pgtok(kstat_delta(oldvm, newvm, "dfree"))
	    / etime);
	adjprintf(" %*ld", 2, pgtok(new->s_sys.ss_deficit));
	adjprintf(" %*.0f", 2, kstat_delta(oldvm, newvm, "scan") / etime);

	(void) snapshot_walk(SNAP_IODEVS, old, new, show_disk, NULL);

	count = df.if_max_iodevs - new->s_nr_iodevs;
	while (count-- > 0)
		adjprintf(" %*d", 2, 0);

	adjprintf(" %*.0f", 4, kstat_delta(oldsys, newsys, "intr") / etime);
	adjprintf(" %*.0f", 4, kstat_delta(oldsys, newsys, "syscall") / etime);
	adjprintf(" %*.0f", 4, kstat_delta(oldsys, newsys, "pswitch") / etime);
	adjprintf(" %*.0f", 2,
	    kstat_delta(oldsys, newsys, "cpu_ticks_user") * percent_factor);
	adjprintf(" %*.0f", 2, kstat_delta(oldsys, newsys, "cpu_ticks_kernel")
	    * percent_factor);
	adjprintf(" %*.0f\n", 2, (kstat_delta(oldsys, newsys, "cpu_ticks_idle")
	    + kstat_delta(oldsys, newsys, "cpu_ticks_wait"))
	    * percent_factor);
	(void) fflush(stdout);
}

/*ARGSUSED*/
static void
print_disk(void *v, void *v2, void *d)
{
	struct iodev_snapshot *iodev = (struct iodev_snapshot *)v2;

	if (iodev == NULL)
		return;

	(void) printf("%c%c ", iodev->is_name[0], iodev->is_name[2]);
}

/* ARGSUSED */
static void
printhdr(int sig)
{
	int i = df.if_max_iodevs - ss->s_nr_iodevs;

	if (sig == SIGCONT)
		caught_cont = 1;

	if (pflag) {
		(void) printf("     memory           page          ");
		(void) printf("executable      anonymous      filesystem \n");
		(void) printf("   swap  free  re  mf  fr  de  sr  ");
		(void) printf("epi  epo  epf  api  apo  apf  fpi  fpo  fpf\n");
		lines = REPRINT;
		return;
	}

	(void) printf(" kthr      memory            page            ");
	(void) printf("disk          faults      cpu\n");

	if (swflag)
		(void) printf(" r b w   swap  free  si  so pi po fr de sr ");
	else
		(void) printf(" r b w   swap  free  re  mf pi po fr de sr ");

	(void) snapshot_walk(SNAP_IODEVS, NULL, ss, print_disk, NULL);

	while (i-- > 0)
		(void) printf("-- ");

	(void) printf("  in   sy   cs us sy id\n");
	lines = REPRINT;
}

static void
sum_out(char const *pretty, kstat_t *ks, char *name)
{
	kstat_named_t *ksn = kstat_data_lookup(ks, name);
	if (ksn == NULL) {
		fail(0, "kstat_data_lookup('%s', '%s') failed",
		    ks->ks_name, name);
	}

	(void) printf("%9llu %s\n", ksn->value.ui64, pretty);
}

static void
zero_out(char const *pretty)
{
	(void) printf("%9u %s\n", 0, pretty);
}

static void
dosum(struct sys_snapshot *ss)
{
	uint64_t total_faults;
	kstat_named_t *ksn;
	long double nchtotal;
	uint64_t nchhits;

	zero_out("swap ins");
	zero_out("swap outs");
	zero_out("pages swapped in");
	zero_out("pages swapped out");

	ksn = kstat_data_lookup(&ss->ss_agg_vm, "hat_fault");
	if (ksn == NULL) {
		fail(0, "kstat_data_lookup('%s', 'hat_fault') failed",
		    ss->ss_agg_vm.ks_name);
	}
	total_faults = ksn->value.ui64;
	ksn = kstat_data_lookup(&ss->ss_agg_vm, "as_fault");
	if (ksn == NULL) {
		fail(0, "kstat_data_lookup('%s', 'as_fault') failed",
		    ss->ss_agg_vm.ks_name);
	}
	total_faults += ksn->value.ui64;

	(void) printf("%9llu total address trans. faults taken\n",
	    total_faults);

	sum_out("page ins", &ss->ss_agg_vm, "pgin");
	sum_out("page outs", &ss->ss_agg_vm, "pgout");
	sum_out("pages paged in", &ss->ss_agg_vm, "pgpgin");
	sum_out("pages paged out", &ss->ss_agg_vm, "pgpgout");
	sum_out("total reclaims", &ss->ss_agg_vm, "pgrec");
	sum_out("reclaims from free list", &ss->ss_agg_vm, "pgfrec");
	sum_out("micro (hat) faults", &ss->ss_agg_vm, "hat_fault");
	sum_out("minor (as) faults", &ss->ss_agg_vm, "as_fault");
	sum_out("major faults", &ss->ss_agg_vm, "maj_fault");
	sum_out("copy-on-write faults", &ss->ss_agg_vm, "cow_fault");
	sum_out("zero fill page faults", &ss->ss_agg_vm, "zfod");
	sum_out("pages examined by the clock daemon", &ss->ss_agg_vm, "scan");
	sum_out("revolutions of the clock hand", &ss->ss_agg_vm, "rev");
	sum_out("pages freed by the clock daemon", &ss->ss_agg_vm, "dfree");
	sum_out("forks", &ss->ss_agg_sys, "sysfork");
	sum_out("vforks", &ss->ss_agg_sys, "sysvfork");
	sum_out("execs", &ss->ss_agg_sys, "sysexec");
	sum_out("cpu context switches", &ss->ss_agg_sys, "pswitch");
	sum_out("device interrupts", &ss->ss_agg_sys, "intr");
	sum_out("traps", &ss->ss_agg_sys, "trap");
	sum_out("system calls", &ss->ss_agg_sys, "syscall");

	nchtotal = (long double) ss->ss_nc.ncs_hits.value.ui64 +
	    (long double) ss->ss_nc.ncs_misses.value.ui64;
	nchhits = ss->ss_nc.ncs_hits.value.ui64;
	(void) printf("%9.0Lf total name lookups (cache hits %.0Lf%%)\n",
	    nchtotal, nchhits / denom(nchtotal) * 100);

	sum_out("user   cpu", &ss->ss_agg_sys, "cpu_ticks_user");
	sum_out("system cpu", &ss->ss_agg_sys, "cpu_ticks_kernel");
	sum_out("idle   cpu", &ss->ss_agg_sys, "cpu_ticks_idle");
	sum_out("wait   cpu", &ss->ss_agg_sys, "cpu_ticks_wait");
}

static void
dointr(struct snapshot *ss)
{
	size_t i;
	ulong_t total = 0;

	(void) printf("interrupt         total     rate\n");
	(void) printf("--------------------------------\n");

	for (i = 0; i < ss->s_nr_intrs; i++) {
		(void) printf("%-12.8s %10lu %8.0f\n",
		    ss->s_intrs[i].is_name, ss->s_intrs[i].is_total,
		    ss->s_intrs[i].is_total / etime);
		total += ss->s_intrs[i].is_total;
	}

	(void) printf("--------------------------------\n");
	(void) printf("Total        %10lu %8.0f\n", total, total / etime);
}

static void
usage(void)
{
	(void) fprintf(stderr,
	    "Usage: vmstat [-ipqsS] [-T d|u] [disk ...] "
	    "[interval [count]]\n");
	exit(1);
}
